### PR TITLE
feat: type checking

### PR DESCRIPTION
**Pull request by simoneb**
Attempt to address #168 by using https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html.

I'm not sure this is worth it really. It helps in some cases, it's a pain to make it work in others. 

There are two main categories of types which are difficult to handle: 

- stubbing with sinon. it's hard to tell the type checker that a stubbed function has its original signature, plus, the signature of the sinon stub (e.g. `<stub>.returns` leads to an error)
- decorators: hard to tell the type checker that a fastify server has certain decorators (`fastify.authenticate` leads to an error)

Happy if someone with more familiarity with the type system and patience than myself wants to have a look.

### How to test

- Open this branch in VS code
- All files with typing errors should be red

### CODE REVIEW

File: src/step-10-hooks/routes/user/index.js [hunk 1]
```diff
@@ -6,6 +6,9 @@ const schema = {
   },
 }
 
+/**
+ * @type {import('fastify').FastifyPluginAsync}
```
Comment (Eomm) [inline src/step-10-hooks/routes/user/index.js:10]: I see an error on the 16th line:

```
onRequest: [fastify.authenticate],
```

File: src/step-11-autoload/routes/user/index.js [hunk 2]
```diff
@@ -6,6 +6,9 @@ const schema = {
   },
 }
 
+/**
+ * @type {import('fastify').FastifyPluginAsync}
```
Comment (Eomm) [inline src/step-11-autoload/routes/user/index.js:10]: Same error on this file: `fastify.authenticate`

**Comment by simoneb** [comment]
@Eomm read the description

**Comment by simoneb** [comment]
One option would be to disable such errors so they're reported as "warnings" and they don't stand out so much. This is basically just a matter of turning off the only setting in jsconfig.json or deleting that file altogether.

**Comment by Eomm** [comment]
I was trying without success:

```
/**
 * @typedef { import('fastify').FastifyInstance<> & { authenticate: function } } FastifyInstanceWithAuthenticate
 * @property { function } authenticate
 */

/**
 * @type { import('fastify').FastifyPluginAsync< {}, FastifyInstanceWithAuthenticate> } fastify
 */
```

It should be a matter of configuring the 2nd FastifyPluginAsync generic interface

**Comment by simoneb** [comment]
> I was trying without success:
> 
> ```
> /**
>  * @typedef { import('fastify').FastifyInstance<> & { authenticate: function } } FastifyInstanceWithAuthenticate
>  * @property { function } authenticate
>  */
> 
> /**
>  * @type { import('fastify').FastifyPluginAsync< {}, FastifyInstanceWithAuthenticate> } fastify
>  */
> ```
> 
> It should be a matter of configuring the 2nd FastifyPluginAsync generic interface

I also tried some options but I believe that you need to extend the types from fastify to do this and I'm not sure if it's doable without typescript.

File: package.json [hunk 3]
```diff
@@ -36,7 +37,8 @@
     "postgrator-cli": "^4.0.0",
     "prettier": "^2.4.1",
     "sinon": "^12.0.1",
-    "tap": "^15.1.1"
+    "tap": "^15.1.1",
+    "typescript": "^4.4.4"
```
Comment (willmendesneto) [inline package.json:41]: `tap` and `typescript` packages should be moved into `devDependencies`. Unless I'm missing something, they won't be shipped in any case we're about to build. Could you move them into `devDependencies`, please?

**Comment by simoneb** [comment]
I'm closing this PR because attempts to make typing work in a non-extremely-simple setup doesn't seem to be possible. The main issues are with:

- plugins which extend the fastify core objects (server, request, reply), where types needs to be manually extended
- mocking frameworks like sinon, which turn plain values (e.g. functions) into "super values" (functions having all the sinon API methods. During tests, when you provide a sinon stub to replace a decorator, you would have to extend fastify types to tell the type system that they have additional properties whose type is a sinon stub. This doesn't seem to be possible with jsdoc

**Pull request closed without merging**